### PR TITLE
fix: ip_port schema type crash

### PR DESCRIPTION
**Pull request by zhongwencool**
Fixes : Return 500 crash when update bind as a bad port.
```
2023-06-16T19:14:01.934300+08:00 [error] input-config:, <<"0.0.0.0:188d">>, #{exception => badarg,field => <<"bind">>,path => "root", reason => failed_to_check_field}
2023-06-16T19:14:01.934449+08:00 [error] crasher: initial call: cowboy_stream_h:request_process/3, pid: <0.1183.0>, registered_name: [], error: {#{exception => badarg,field => <<"bind">>,path => "root",reason => failed_to_check_field},[{erlang,list_to_integer,["188d"],[{error_info,#{module => erl_erts_errors}}]},{emqx_schema,to_ip_port,1,[{file,"emqx_schema.erl"},{line,2530}]},{hocon_schema_builtin,convert,2,[{file,"hocon_schema_builtin.erl"},{line,58}]},{hocon_tconf,map_field,4,[{file,"hocon_tconf.erl"},{line,651}]},{hocon_tconf,do_map_union,4,[{file,"hocon_tconf.erl"},{line,784}]},{hocon_tconf,map_field,4,[{file,"hocon_tconf.erl"},{line,599}]},{hocon_tconf,map_one_field_non_hidden,4,[{file,"hocon_tconf.erl"},{line,492}]},{hocon_tconf,map_fields_cont,4,[{file,"hocon_tconf.erl"},{line,431}]},{hocon_tconf,do_map_union,4,[{file,"hocon_tconf.erl"},{line,784}]},{hocon_tconf,map_field,4,[{file,"hocon_tconf.erl"},{line,599}]},{hocon_tconf,map_one_field_non_hidden,4,[{file,"hocon_tconf.erl"},{line,492}]},{hocon_tconf,map_fields_cont,4,[{file,"hocon_tconf.erl"},{line,431}]},{hocon_tconf,map,4,[{file,"hocon_tconf.erl"},{line,301}]},{hocon_tconf,do_check,4,[{file,"hocon_tconf.erl"},{line,267}]},{emqx_dashboard_swagger,check_only,3,[{file,"emqx_dashboard_swagger.erl"},{line,317}]},{emqx_dashboard_swagger,check_request_body,5,[{file,"emqx_dashboard_swagger.erl"},{line,421}]},{emqx_dashboard_swagger,translate_req,3,[{file,"emqx_dashboard_swagger.erl"},{line,305}]},{minirest_handler,do_filter,3,[{file,"/Users/mac/github/emqx/_build/default/lib/minirest/src/minirest_handler.erl"},{line,94}]},{minirest_handler,handle,2,[{file,"/Users/mac/github/emqx/_build/default/lib/minirest/src/minirest_handler.erl"},{line,44}]},{minirest_handler,init,2,[{file,"/Users/mac/github/emqx/_build/default/lib/minirest/src/minirest_handler.erl"},{line,27}]},{cowboy_handler,execute,2,[{file,"cowboy_handler.erl"},{line,41}]},{cowboy_stream_h,execute,3,[{file,"cowboy_stream_h.erl"},{line,318}]},{cowboy_stream_h,request_process,3,[{file,"cowboy_stream_h.erl"},{line,302}]}]}, ancestors: [<0.1182.0>,<0.850.0>,<0.849.0>,ranch_sup,<0.418.0>], message_queue_len: 0, messages: [], links: [<0.1182.0>], dictionary: [], trap_exit: false, status: running, heap_size: 75113, stack_size: 28, reductions: 170532; neighbours:
```
No crash log after this PR.
image
image
image

The `emqx_schema:ip_port()` is support `integer()` format, so  we remove the union type.

https://github.com/emqx/hocon/pull/259

### <samp>🤖 Generated by Copilot at 1f2e018</samp>

This pull request standardizes the schema and validation of listener ports across different modules. It removes the possibility of using integers as ports and replaces it with the `ip_port()` type. It also refactors and adds some helper functions in `emqx_schema.erl`.

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [ ] Added tests for the changes
- [ ] Changed lines covered in coverage report
- [ ] Change log has been added to `changes/{ce,ee}/(feat|perf|fix)-<PR-id>.en.md` files
- [ ] For internal contributor: there is a jira ticket to track this change
- [ ] If there should be document changes, a PR to emqx-docs.git is sent, or a jira ticket is created to follow up
- [ ] Schema changes are backward compatible

## Checklist for CI (.github/workflows) changes

- [ ] If changed package build workflow, pass [this action](https://github.com/emqx/emqx/actions/workflows/build_packages.yaml) (manual trigger)
- [ ] Change log has been added to `changes/` dir for user-facing artifacts update

### CODE REVIEW

**Review by thalesmg — APPROVED** [review]

**Review by HJianBo — DISMISSED** [review]

File: apps/emqx/src/emqx_schema.erl [hunk 1]
```diff
@@ -1758,7 +1758,7 @@ base_listener(Bind) ->
             )},
         {"bind",
             sc(
-                hoconsc:union([ip_port(), integer()]),
+                ip_port(),
                 #{
```
Comment (zhongwencool) [inline apps/emqx/src/emqx_schema.erl:1762]: Remove integer: ip_port() also support setting as `1883`.

File: apps/emqx/src/emqx_schema.erl [hunk 2]
```diff
@@ -2418,21 +2418,21 @@ mk_duration(Desc, OverrideMeta) ->
 to_duration(Str) ->
     case hocon_postprocess:duration(Str) of
         I when is_integer(I) -> {ok, I};
-        _ -> {error, Str}
```
Comment (zhongwencool) [inline apps/emqx/src/emqx_schema.erl:2421]: hocon 0.39.9 has fixed a issue: `typerelf:from_string` conversion is not executed when value is integer  https://github.com/emqx/hocon/pull/259
In order to be compatible with the previous format.
```
%% {error, "1"} = hocon_postprocess:duration("1")
%% we should ensure:
to_duration("12") -> 12.
to_duration_s("12") -> 12s
to_duration_ms("12") -> 12ms

to_timeout_duration("12") -> 12
to_timeout_duration_ms("12") ->  12ms
to_timeout_duration_s("12") -> 12s
```

File: apps/emqx/src/emqx_schema.erl [hunk 3]
```diff
@@ -2483,6 +2483,13 @@ to_wordsize(Str) ->
         Error -> Error
     end.
 
+to_integer(Str) when is_list(Str) ->
+    case string:to_integer(Str) of
+        {Int, []} -> {ok, Int};
+        {Int, <<>>} -> {ok, Int};
```
[the remark this answers is not in the document]
Comment (zhongwencool) [reply apps/emqx/src/emqx_schema.erl:2489]: Good catch, I will remove is_list

**Review by HJianBo — APPROVED** [review]
LGTM

**Comment by id** [comment]
Do we need a changelog entry for this one @zhongwencool ?

**Comment by zhongwencool** [comment]
> Do we need a changelog entry for this one @zhongwencool ?

@id OK， I add changelog for this PR https://github.com/emqx/emqx/pull/11192

**Pull request merged**